app/app.go

```diff
@@ -450,3 +450,42 @@ func (a *App) Handle404(w http.ResponseWriter, r *http.Request) {
 
 	utils.RenderWebError(err, w, r)
 }
+
+// This function migrates the default built in roles from code/config to the database.
+func (a *App) DoAdvancedPermissionsMigration() {
+	// If the migration is already marked as completed, don't do it again.
+	if result := <-a.Srv.Store.System().GetByName(ADVANCED_PERMISSIONS_MIGRATION_KEY); result.Err == nil {
+		return
+	}
+
+	l4g.Info("Migrating roles to database.")
+	roles := model.MakeDefaultRoles()
+	roles = utils.SetRolePermissionsFromConfig(roles, a.Config())
+
+	allSucceeded := true
+
+	for _, role := range roles {
+		if result := <-a.Srv.Store.Role().Save(role); result.Err != nil {
```

Comment:
I'm 2/5 here, but It can leave the database in an inconsistent databases state. Could have sense to wrap it in a database transaction?

Reply:
Just wondering, which part would you wrap in a transaction if you were doing this? All the "saves" in the loop, and then rolling back if any of them fail?

Reply:
Yes, exactly, all the migration (roles update/creation) is applied or none is applied. Actually, for me, it must also include the migration confirmation stored in the database.

Reply:
My expected behavior here is, If you can run the transaction, keep going, if not, fail with an error exit the application. That is valid with the idempotent approach to, but If i want to migrate a database, I expect that the migration is all or nothing, because if I want to rollback to the previous version because something is failing in the migration, I don't want my data in an inconsistent state.

Said that, I understand the idempotent approach is good enough, and of course, the transactional approach have it owns problems, and implies a lot of extra effort.

Reply:
OK I'm going to think about this some more, and see if anyone else has opinions.

Reply:
I agree that a transaction would be ideal here. If the logs tell me that the migration completed successfully, then I expect that all users have the permissions that I intended them to have, rather than some users being stuck in a fail safe mode. I think migrations should be all or nothing, but this is something that I'm comfortable pushing off until later for the sake of getting this merged if we create a JIRA ticket to track it.

Reply:
I would be much happier with this if we just ensured that stale roles from a previously failed attempt never have any effect. After the `GetByName` call below, make sure the existing role actually matches what we expect. If it doesn't, update it. Easy.

If we do that, I don't think any other concurrency control is necessary.

Alternatively, maybe you could go in a completely different direction and avoid the need for a migration altogether by treating the database roles as overrides? If one isn't present in the database, just use the default.

Reply:
Implemented your first suggestion.